plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/use-form-validation.ts

```diff
@@ -0,0 +1,269 @@
+			break;
+	}
+
+	const partialSchema = formFields.reduce( ( acc, field ) => {
+		if (
+			! field.hidden &&
+			typeof field.rules?.validation === 'object' &&
+			! Array.isArray( field.rules.validation ) &&
+			( field.required || values[ field.key as keyof typeof values ] )
+		) {
+			acc[ field.key ] = field.rules.validation;
+		}
+		return acc;
+	}, {} as Record< keyof FormFields, JSONSchemaType< DocumentObject< typeof formType > > > );
+
+	let schemaErrorsMap = EMPTY_OBJECT;
+
+	if ( Object.keys( partialSchema ).length > 0 && parser ) {
+		const schema = {
+			type: 'object',
+			properties: {},
+		};
+		switch ( formType ) {
+			case 'shipping':
+				schema.properties = {
+					customer: {
+						type: 'object',
+						properties: {
+							shipping_address: {
+								type: 'object',
+								properties: partialSchema,
+							},
+						},
+					},
+				};
+				break;
+			case 'billing':
+				schema.properties = {
+					customer: {
+						type: 'object',
+						properties: {
+							billing_address: {
+								type: 'object',
+								properties: partialSchema,
+							},
+						},
+					},
+				};
+				break;
+			default:
+				schema.properties = {
+					checkout: {
+						type: 'object',
+						properties: {
+							additional_fields: {
+								type: 'object',
+								properties: partialSchema,
+							},
+						},
+					},
+				};
+				break;
+		}
+		const validate = parser.compile( schema );
+		const result = validate( data );
+
+		if ( ! result && validate.errors ) {
+			schemaErrorsMap = getErrorsMap( validate.errors, formFields );
+		} else {
+			schemaErrorsMap = EMPTY_OBJECT;
+		}
+	}
+
+	const customValidation =
+		values &&
+		formFields
+			.map( ( field ) => {
+				if ( schemaErrorsMap[ field.key ] ) {
+					return [ field.key, schemaErrorsMap[ field.key ] ];
+				}
+
+				// Pass validation if the field is not required and is empty.
+				if (
+					! field.required &&
+					! values[ field.key as keyof typeof values ]
+				) {
+					return null;
+				}
+
+				if (
+					field.key === 'postcode' &&
+					'country' in values &&
+					! isPostcode( {
+						postcode: values.postcode,
+						country: values.country,
+					} )
+				) {
+					return [
+						field.key,
+						__( 'Please enter a valid postcode', 'woocommerce' ),
+					];
+				}
+
+				if (
+					field.key === 'email' &&
+					'email' in values &&
+					! isEmail( values.email )
+				) {
+					return [
+						field.key,
+						__(
+							'Please enter a valid email address',
+							'woocommerce'
+						),
+					];
+				}
+
+				return null;
+			} )
+			.filter( Boolean );
+
+	if (
+		! fastDeepEqual(
+			currentResults.current,
+			Object.fromEntries(
+				customValidation as [ keyof FormFields, string ][]
+			)
+		)
+	) {
+		currentResults.current = Object.fromEntries(
+			customValidation as [ keyof FormFields, string ][]
+		) as Record< keyof FormFields, string >;
+	}
+
+	return {
+		errors: currentResults.current,
+		previousErrors,
```

Comment:
Should we return `previousErrors` here or allow the consumer of this hook to do it if the use case demands it?